alacritty/src/wayland_theme.rs

```diff
@@ -32,51 +34,61 @@ impl AlacrittyWaylandTheme {
-            },
-        }
-    }
-}
-
-impl WaylandTheme for AlacrittyWaylandTheme {
-    fn primary_color(&self, _window_active: bool) -> [u8; 4] {
-        [0xff, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn secondary_color(&self, window_active: bool) -> [u8; 4] {
+    fn color_icon_color(&self, color: Rgb, status: ButtonState, window_active: bool) -> Rgb {
         if window_active {
-            [0xff, self.foreground.r, self.foreground.g, self.foreground.b]
+            match status {
+                ButtonState::Hovered => color,
+                ButtonState::Idle => self.foreground,
+                ButtonState::Disabled => self.dim_foreground,
+            }
         } else {
-            [0xff, self.dim_foreground.r, self.dim_foreground.g, self.dim_foreground.b]
+            self.dim_foreground
         }
     }
+}
 
-    fn close_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn close_button_icon_color(&self, status: ButtonState) -> [u8; 4] {
-        self.color_icon_color(self.hovered_close_icon, status)
-    }
+impl WaylandTheme for AlacrittyWaylandTheme {
+    fn element_color(&self, element: Element, window_active: bool) -> ARGBColor {
+        let Rgb { r, g, b } = match element {
+            Element::Bar => self.background,
+            Element::Separator => self.background,
+            Element::Text => {
+                if window_active {
+                    self.foreground
+                } else {
+                    self.dim_foreground
+                }
+            },
+        };
 
-    fn maximize_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
+        ARGBColor { a: 0xff, r, g, b }
     }
 
-    fn maximize_button_icon_color(&self, status: ButtonState) -> [u8; 4] {
-        self.color_icon_color(self.hovered_maximize_icon, status)
-    }
+    fn button_color(
+        &self,
+        button: Button,
+        state: ButtonState,
+        foreground: bool,
+        window_active: bool,
+    ) -> ARGBColor {
+        let (a, Rgb { r, g, b }) = if foreground {
+            let color = match button {
+                Button::Minimize => {
+                    self.color_icon_color(self.hovered_minimize_icon, state, window_active)
+                },
+                Button::Maximize => {
+                    self.color_icon_color(self.hovered_maximize_icon, state, window_active)
+                },
+                Button::Close => {
+                    self.color_icon_color(self.hovered_close_icon, state, window_active)
+                },
+            };
 
-    fn minimize_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
-    }
+            (0xff, color)
```

Comment:
Converting the color to ARGBColor here seems far easier than destructuring it and then creating it as return value, especially if the other branch is changed.